File: backend/domain/conversation/message/entity/message.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	model "github.com/coze-dev/coze-studio/backend/crossdomain/message/model"
)

type Message = model.Message

type ListMeta struct {
	ConversationID int64                `json:"conversation_id"`
	RunID          []*int64             `json:"run_id"`
	UserID         string               `json:"user_id"`
	AgentID        int64                `json:"agent_id"`
	OrderBy        *string              `json:"order_by"`
	Limit          int                  `json:"limit"`
	Cursor         int64                `json:"cursor"`    // message id
	Direction      ScrollPageDirection  `json:"direction"` //  "prev" "Next"
	MessageType    []*model.MessageType `json:"message_type"`
}

type ListResult struct {
	Messages   []*Message          `json:"messages"`
	PrevCursor int64               `json:"prev_cursor"`
	NextCursor int64               `json:"next_cursor"`
	HasMore    bool                `json:"has_more"`
	Direction  ScrollPageDirection `json:"direction"`
}

type GetByRunIDsRequest struct {
	ConversationID int64   `json:"conversation_id"`
	RunID          []int64 `json:"run_id"`
}

type DeleteMeta struct {
	ConversationID *int64  `json:"conversation_id"`
	MessageIDs     []int64 `json:"message_ids"`
	RunIDs         []int64 `json:"run_ids"`
}

type BrokenMeta struct {
	ID       int64  `json:"id"`
	Position *int32 `json:"position"`
}

File: frontend/packages/common/chat-area/plugin-message-grab/src/services/life-cycle/message.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getOriginContentText } from '@coze-common/text-grab';
import {
  ContentType,
  type OnBeforeAppendSenderMessageIntoStore,
  WriteableMessageLifeCycleService,
  type OnAfterAppendSenderMessageIntoStore,
} from '@coze-common/chat-area';

import {
  EventNames,
  type GrabPluginBizContext,
} from '../../types/plugin-biz-context';

export class GrabMessageLifeCycleService extends WriteableMessageLifeCycleService<GrabPluginBizContext> {
  onBeforeAppendSenderMessageIntoStore(
    ctx: OnBeforeAppendSenderMessageIntoStore,
  ) {
    const { quoteContent, updateQuoteContentMapByImmer } =
      this.pluginInstance.pluginBizContext.storeSet.useQuoteStore.getState();

    if (!quoteContent || ctx.from === 'shortcut') {
      return ctx;
    }

    const originMessage = ctx.message;

    const newContent = {
      item_list: [
        {
          type: 'text',
          text: originMessage.content,
        },
      ],
      refer_items: [] as unknown[],
    };

    newContent.refer_items.push({
      type: ContentType.Text,
      text: getOriginContentText(quoteContent),
    });

    const newMessage = {
      ...originMessage,
      content_type: ContentType.Mix,
      content: JSON.stringify(newContent),
      content_obj: newContent,
    };

    updateQuoteContentMapByImmer(quoteContentMap => {
      const localMessageId = newMessage.extra_info.local_message_id;
      if (!quoteContentMap[localMessageId]) {
        quoteContentMap[localMessageId] = quoteContent;
      }
    });

    return {
      ...ctx,
      message: newMessage,
    };
  }

  onAfterAppendSenderMessageIntoStore(
    ctx: OnAfterAppendSenderMessageIntoStore,
  ) {
    if (ctx.from === 'shortcut') {
      return;
    }

    const { storeSet } = this.pluginInstance.pluginBizContext;
    const { updateQuoteVisible, updateQuoteContent } =
      storeSet.useQuoteStore.getState();
    updateQuoteVisible(false);
    updateQuoteContent(null);

    return;
  }

  onAfterProcessReceiveMessage() {
    const { emit } = this.pluginInstance.pluginBizContext.eventCenter;
    emit(EventNames.OnMessageUpdate);
  }
}
